### PR TITLE
feat: add deno package

### PR DESCRIPTION
**Pull request by jacob-ebey**
This PR adds a server adapter for Deno. Currently the approach to building a Remix app on Deno will be the same "bundle everything" approach that we are taking with CF.

Deno has "legacy" support for bare module imports from node_modules, but this is not in line with the spirit of native URL imports Deno is great at, and not all Deno deployment targets support it. I will be re-visiting import_map support, both bundling and externalized. This will require more thought as it must support a good UX and DX for Deno *and* browsers.

### CODE REVIEW

**Comment by mcansh** [comment]
i think we want/need to add this to the `workspaces` in the root package.json too

File: packages/remix-deno/server.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  }
+}: {
+  build: ServerBuild;
+  mode?: string;
+  staticFiles?: {
+    cacheControl?: string | ((url: URL) => string);
+    publicDir?: string;
+  };
+}) {
+  let remixHandler = createRequestHandler(build, {}, mode);
```
Comment (mjackson) [inline packages/remix-deno/server.ts:29]: Let's split this up into 3 separate functions:

- `createRequestHandler`
- `createRequestHandlerWithStaticFiles`
- `serveStaticFiles`

File: packages/remix-deno/server.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
+import { createRequestHandler } from "@remix-run/server-runtime";
+import type { ServerBuild } from "@remix-run/server-runtime";
+
+import mime from "mime";
+
+export function defaultCacheControl(url: URL) {
+  if (url.pathname.startsWith("/build/")) {
```
Comment (mjackson) [inline packages/remix-deno/server.ts:7]: The `/build/` should be an arg we get from the user.

File: packages/remix-deno/server.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        headers.set("Cache-Control", cacheControl(url));
+      } else if (cacheControl) {
+        headers.set("Cache-Control", cacheControl);
+      } else {
+        headers.set("Cache-Control", defaultCacheControl(url));
+      }
+
+      // @ts-expect-error
+      const file = await Deno.readFile(
+        `${publicDir || "./public"}${url.pathname}`
```
Comment (mjackson) [inline packages/remix-deno/server.ts:51]: Let's use something like `path.join` here to resolve the file path.

**Review by mjackson — CHANGES_REQUESTED** [review]

File: packages/remix-deno/server.ts [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+  return new Response(file, { headers });
+}
+
+export function createRequestHandlerWithStaticFiles<Context = unknown>({
+  build,
+  mode,
+  getLoadContext,
+  staticFiles = {
+    publicDir: "./public",
```
Comment (mjackson) [inline packages/remix-deno/server.ts:80]: Should we have a `cacheControl` option here as well to match the args of `serveStaticFiles`?

File: packages/remix-react/package.json [hunk 5]
```diff
@@ -14,6 +14,7 @@
   "main": "index.js",
   "module": "esm/index.js",
   "dependencies": {
+    "history": "^5.2.0",
```
Comment (mjackson) [inline packages/remix-react/package.json:17]: Is this just something that was missing? I thought we didn't need an explicit dep on `history` now...

File: jest.config.js [hunk 6]
```diff
@@ -11,6 +11,19 @@ module.exports = {
       testEnvironment: "node",
       testMatch: ["<rootDir>/packages/create-remix/**/*-test.[jt]s?(x)"]
     },
+    {
+      displayName: "deno",
+      testEnvironment: "jsdom",
+      testMatch: ["<rootDir>/packages/remix-deno/**/*-test.[jt]s?(x)"],
+      moduleNameMapper: {
+        "https://deno.land/std/path/mod.ts":
+          "<rootDir>/packages/remix-deno/__tests__/pathmock.js"
```
Comment (mjackson) [inline jest.config.js:20]: Clever 👍

File: packages/remix-deno/server.ts [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    cacheControl,
+    publicDir = "./public",
+    assetsPublicPath = "/build/"
+  }: {
+    cacheControl?: string | ((url: URL) => string);
+    publicDir?: string;
+    assetsPublicPath?: string;
+  }
+) {
+  const url = new URL(request.url);
```
Comment (mjackson) [inline packages/remix-deno/server.ts:53]: nit: use `let` instead of `const`

File: packages/remix-react/package.json (hunk 5 above)
Comment (jacob-ebey) [reply packages/remix-react/package.json:17]: Whenever we import from a package in code it should be listed in deps regardless of if it's a dep of another dep we have listed. It helps bundlers and ESM CDN's like esm.sh. Without it more config is required to get remote modules to load. This is currently the only thing stopping us from consuming @remix-run packages from esm.sh.

File: packages/remix-deno/server.ts (hunk 4 above)
Comment (jacob-ebey) [reply packages/remix-deno/server.ts:80]: It defaults later on in the codepath in the serveStaticFiles if it's not provided :) No need to add it here.

File: packages/remix-react/package.json (hunk 5 above)
Comment (mjackson) [reply packages/remix-react/package.json:17]: Yeah, I get that. I guess what I'm saying is let's remove whatever imports we're using from `history` in `remix-react` so we don't have to do this

**Review by mjackson — APPROVED** [review]

File: packages/remix-react/package.json (hunk 5 above)
Comment (jacob-ebey) [reply packages/remix-react/package.json:17]: `parsePath` and `history` need to be re-exported before we can do this.

**Pull request merged**